LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/UdpServer.cs

```diff
@@ -76,36 +78,86 @@ async Task RunUdpListener()
             {
                 UdpReceiveResult receivedResults = await udpClient.ReceiveAsync();
 
-                //Logger.Log($"UDP message received ({receivedResults.Buffer.Length} bytes) from port: {receivedResults.RemoteEndPoint.Port}");
+               //Logger.Log($"UDP message received ({receivedResults.Buffer[3]}) from port: {receivedResults.RemoteEndPoint.Port} and IP: {receivedResults.RemoteEndPoint.Address.ToString()}",LoggingLevel.Always);
+                 
 
-
-                //Todo check that is an ack only, we could do a better check in a future verstion
-                if (receivedResults.Buffer.Length == 12)
+                switch (receivedResults.Buffer[3])
```

Comment:
I have a few suggestions to, maybe, improve code readability and chance of running into unhandled exceptions.

Move all the byte manipulation to classes that already do that, namely PhysicalPayload. Moreover, accessing receivedResults[3] without checks will kill the process if it is null or has length < 4.

As suggestion, to improving the readability, I would make a few changes to the code to not rely on magic numbers (`switch buffer[3]`, `case 0x02`), specially that you already have an enum defining these values.

* Move low level manipulation to PhysicalPayload class. 

```C#

public enum PhysicalIdentifier : byte // <- use byte here since you know it has to fit a into a byte
{
        PUSH_DATA, PUSH_ACK, PULL_DATA, PULL_RESP, PULL_ACK, TX_ACK, 
        UNKOWN = byte.MaxValue // <-- can become a problem in future if they ever use 255 to define a message type
}

public class PhysicalPayload
{
        // Position in payload where the identifier is located (3)
        const int PHYSICAL_IDENTIFIER_INDEX = 3;

        // Gets the identifier of the payload
        public static PhysicalIdentifier GetIdentifierFromPayload(byte[] packet)
        {
            // Unknown if: packet is null or does not have the physical identifier byte
            if (packet == null || packet.Length < (PHYSICAL_IDENTIFIER_INDEX + 1))
                return PhysicalIdentifier.UNKOWN;

            return (PhysicalIdentifier)packet[PHYSICAL_IDENTIFIER_INDEX];
        }
  
         ...
}
```

* Clear UdpServer.cs a bit. Don't forget to check the case where buffer.Length < 12 on the ProcessTxAck method.

```C#
 async Task RunUdpListener()
 {


            IPEndPoint endPoint = new IPEndPoint(IPAddress.Any, PORT);
            udpClient = new UdpClient(endPoint);

            Logger.Log($"LoRaWAN server started on port {PORT}", Logger.LoggingLevel.Always);


            while (true)
            {
                UdpReceiveResult receivedResults = await udpClient.ReceiveAsync();

               //Logger.Log($"UDP message received ({receivedResults.Buffer[3]}) from port: {receivedResults.RemoteEndPoint.Port} and IP: {receivedResults.RemoteEndPoint.Address.ToString()}",LoggingLevel.Always);
                
                switch (PhysicalPayload.GetIdentifierFromPayload(receivedResults.Buffer))
                {
                    case PhysicalIdentifier.PULL_DATA:
                    {
                        //In this case we have a keep-alive PULL_DATA packet we don't need to start the engine and can return immediately a response to the challenge                                            
                        if (pullAckRemoteLoRaAggregatorPort == 0)
                        {
                            remoteLoRaAggregatorIp = receivedResults.RemoteEndPoint.Address;
                            pullAckRemoteLoRaAggregatorPort = receivedResults.RemoteEndPoint.Port;
                        }
                        SendAcknowledgementMessage(receivedResults, PhysicalIdentifier.PUSH_DATA, pullAckRemoteLoRaAggregatorPort);
                        break;
                    }


                    
                    case PhysicalIdentifier.PUSH_DATA:
                    {
                        //This is a PUSH_DATA (upstream message).
                        if (pushDataRemoteLoRaAggregatorPort == 0)
                        {
                            remoteLoRaAggregatorIp = receivedResults.RemoteEndPoint.Address;
                            pushDataRemoteLoRaAggregatorPort = receivedResults.RemoteEndPoint.Port;
                        }
                        SendAcknowledgementMessage(receivedResults, PhysicalIdentifier.PUSH_ACK, pushDataRemoteLoRaAggregatorPort);

                        // Message processing runs in the background
                        MessageProcessor messageProcessor = new MessageProcessor(this.configuration, this.loraDeviceInfoManager);

                        _ = Task.Run(async () => {
                            try
                            {
                                var resultMessage = await messageProcessor.ProcessMessageAsync(receivedResults.Buffer);
                                if (resultMessage != null && resultMessage.Length > 0)
                                    await this.UdpSendMessage(resultMessage, remoteLoRaAggregatorIp.ToString(), pullAckRemoteLoRaAggregatorPort);
                            }
                            catch (Exception ex)
                            {
                                Logger.Log($"Error processing the message {ex.Message}, {ex.StackTrace}", Logger.LoggingLevel.Error);
                            }
                        });

                        break;
                    }

                    
                    case PhysicalIdentifier.TX_ACK:
                    {
                        //This is a ack to a transmission we did previously
                        ProcessTxAck(receivedResults.Buffer);
                        break;
                    }

                    default:
                    {
                        Logger.Log("Unknown packet type being received", LoggingLevel.Error);
                        break;
                     }
                }

            }
        }

        private void ProcessTxAck(byte[] data)
        {            
            if (data == null || data.Length != 12)
            {
                Logger.Log(String.Format("Packet with id {0} had a problem to be transmitted over the air :{1}",
                ConversionHelper.ByteArrayToString(data.RangeSubset(1, 2)),

                // TODO: MIK: this will throw an exception if data.Length < 12
                Encoding.UTF8.GetString(data.RangeSubset(12, data.Length - 12)))
                , LoggingLevel.Error);
            }
            else
            {
                    Logger.Log(String.Format("Packet with id {0} successfully transmitted by the aggregator",
                ConversionHelper.ByteArrayToString(data.RangeSubset(1, 2)))
                , LoggingLevel.Full);
            }
        }

        private void SendAcknowledgementMessage(UdpReceiveResult receivedResults, PhysicalIdentifier messageType, int remotePort)
        {
            // OPTIONAL: Use PhysicalPayload as the single logical place where low level messages are built:
            // var response = PhysicalPayload.CreateAck(receivedResults.Buffer, messageType);

            byte[] response = new byte[4]{
                            receivedResults.Buffer[0],
                            receivedResults.Buffer[1],
                            receivedResults.Buffer[2],
                            (byte)messageType
                            };
            _ = UdpSendMessage(response, remoteLoRaAggregatorIp.ToString(), remotePort);
        }
```